src/rpc/core_rpc_server.cpp

```diff
@@ -2078,10 +2078,23 @@ namespace cryptonote
   //------------------------------------------------------------------------------------------------------------------------------
   bool core_rpc_server::on_get_output_distribution(const COMMAND_RPC_GET_OUTPUT_DISTRIBUTION::request& req, COMMAND_RPC_GET_OUTPUT_DISTRIBUTION::response& res, epee::json_rpc::error& error_resp)
   {
+    PERF_TIMER(on_get_output_distribution);
     try
     {
       for (uint64_t amount: req.amounts)
       {
+        static boost::mutex mutex;
```

Comment:
Is this a micro optimization that makes the code more complicated while there's also some multi-second stuff happening while the mutex is lock ?

Reply:
Is this what you had in mind ?

```
diff --git a/src/rpc/core_rpc_server.cpp b/src/rpc/core_rpc_server.cpp
index 443d8c8..483d50c 100644
--- a/src/rpc/core_rpc_server.cpp
+++ b/src/rpc/core_rpc_server.cpp
@@ -2083,15 +2083,19 @@ namespace cryptonote
     {
       for (uint64_t amount: req.amounts)
       {
-        static boost::mutex mutex;
-        boost::unique_lock<boost::mutex> lock(mutex);
-        static std::vector<uint64_t> cached_distribution;
-        static uint64_t cached_from = 0, cached_to = 0, cached_start_height = 0, cached_base = 0;
-        static bool cached = false;
-
-        if (cached && amount == 0 && cached_from == req.from_height && cached_to == req.to_height)
+        static struct D
+        {
+          boost::mutex mutex;
+          std::vector<uint64_t> cached_distribution;
+          uint64_t cached_from, cached_to, cached_start_height, cached_base;
+          bool cached;
+          D(): cached_from(0), cached_to(0), cached_start_height(0), cached_base(0), cached(false) {}
+        } d;
+        boost::unique_lock<boost::mutex> lock(d.mutex);
+
+        if (d.cached && amount == 0 && d.cached_from == req.from_height && d.cached_to == req.to_height)
         {
-          res.distributions.push_back({amount, cached_start_height, cached_distribution, cached_base});
+          res.distributions.push_back({amount, d.cached_start_height, d.cached_distribution, d.cached_base});
           continue;
         }
 
@@ -2118,12 +2122,12 @@ namespace cryptonote
 
         if (amount == 0)
         {
-          cached_from = req.from_height;
-          cached_to = req.to_height;
-          cached_distribution = distribution;
-          cached_start_height = start_height;
-          cached_base = base;
-          cached = true;
+          d.cached_from = req.from_height;
+          d.cached_to = req.to_height;
+          d.cached_distribution = distribution;
+          d.cached_start_height = start_height;
+          d.cached_base = base;
+          d.cached = true;
         }
 
         res.distributions.push_back({amount, start_height, std::move(distribution), base});
```

Reply:
Yes this is a micro-optimization. Each **function-local** static variable has to be initialized at compile-time, or if not possible, synchronized on first use at run-time. `boost::mutex` and `std::vector` are not marked as `constexpr`, and therefore are more likely to require separate runtime initialization (double-check lock), particularly because the current code has a lock acquisition between the statics. The compiler could probably legally batch the static creation double-check locking, but probably not with that lock in between. Putting them into a struct just makes it easy to declare that all of these need to be/can be initialized at once.

Also `std::mutex` is marked as `constexpr`, which is one of its advantages over boost right now.

Reply:
std mutexes (mutices ?) break on some platform, so we only use boost::mutex anyway.